File: database/schemas/route.js
// Route Schema for MongoDB
// Collection: routes

{
  "_id": ObjectId,
  "operatorId": ObjectId, // Reference to users collection
  "busId": ObjectId, // Reference to buses collection
  "from": {
    "city": String, // Required
    "state": String,
    "location": String, // Bus stand/stop name
    "coordinates": {
      "lat": Number,
      "lng": Number
    }
  },
  "to": {
    "city": String, // Required
    "state": String,
    "location": String,
    "coordinates": {
      "lat": Number,
      "lng": Number
    }
  },
  "via": [{
    "city": String,
    "location": String,
    "coordinates": {
      "lat": Number,
      "lng": Number
    },
    "stopTime": Number // Minutes to wait
  }],
  "distance": Number, // Kilometers
  "duration": Number, // Minutes
  "departureTime": String, // HH:mm format
  "arrivalTime": String, // HH:mm format
  "daysOfWeek": [Number], // 0=Sunday, 1=Monday, etc.
  "basePrice": Number, // Base fare
  "dynamicPricing": {
    "enabled": Boolean,
    "minPrice": Number,
    "maxPrice": Number,
    "currentMultiplier": Number // 1.0 = base price
  },
  "isActive": Boolean, // Default: true
  "totalBookings": Number, // Default: 0
  "averageRating": Number, // Default: 0
  "createdAt": Date,
  "updatedAt": Date
}

// Indexes
db.routes.createIndex({ "operatorId": 1 });
db.routes.createIndex({ "busId": 1 });
db.routes.createIndex({ "from.city": 1, "to.city": 1 });
db.routes.createIndex({ "isActive": 1 });
db.routes.createIndex({ "departureTime": 1 });
db.routes.createIndex({ "daysOfWeek": 1 });

File: src/main/java/com/busberry/model/Route.java
package com.busberry.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.List;

@Document(collection = "routes")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Route {

    @Id
    private String id;

    @Indexed
    private String operatorId;

    @Indexed
    private String busId;

    private Location from;
    private Location to;
    private List<ViaStop> via;

    private Double distance; // Kilometers
    private Integer duration; // Minutes

    private String departureTime; // HH:mm format
    private String arrivalTime; // HH:mm format
    private List<Integer> daysOfWeek; // 0=Sunday, 1=Monday, etc.

    private Double basePrice;
    private DynamicPricing dynamicPricing;

    @Builder.Default
    private Boolean isActive = true;

    @Builder.Default
    private Integer totalBookings = 0;

    @Builder.Default
    private Double averageRating = 0.0;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class Location {
        private String city;
        private String state;
        private String location; // Bus stand/stop name
        private Coordinates coordinates;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class Coordinates {
        private Double lat;
        private Double lng;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class ViaStop {
        private String city;
        private String location;
        private Coordinates coordinates;
        private Integer stopTime; // Minutes to wait
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class DynamicPricing {
        @Builder.Default
        private Boolean enabled = false;
        private Double minPrice;
        private Double maxPrice;
        @Builder.Default
        private Double currentMultiplier = 1.0;
    }
}

